src/coreclr/vm/callstubgenerator.cpp

```diff
@@ -1193,6 +1193,142 @@ CallStubHeader *CallStubGenerator::GenerateCallStub(MethodDesc *pMD, AllocMemTra
+
+    LoaderAllocator *pLoaderAllocator = pMD->GetLoaderAllocator();
+    S_SIZE_T finalStubSize(sizeof(CallStubHeader) + m_routineIndex * sizeof(PCODE));
+    void *pHeaderStorage = pamTracker->Track(pLoaderAllocator->GetHighFrequencyHeap()->AllocMem(finalStubSize));
+
+    CallStubHeader *pHeader = new (pHeaderStorage) CallStubHeader(m_routineIndex, pRoutines, ALIGN_UP(m_totalStackSize, STACK_ALIGN_SIZE), m_pInvokeFunction);
+
+    return pHeader;
+}
+
+struct CachedCallStubKey
+{
+    CachedCallStubKey(int32_t hashCode, int numRoutines, PCODE *pRoutines, int totalStackSize, CallStubHeader::InvokeFunctionPtr pInvokeFunction)
+     : HashCode(hashCode), NumRoutines(numRoutines), TotalStackSize(totalStackSize), Invoke(pInvokeFunction), Routines(pRoutines)
+    {
+    }
+
+    bool operator==(const CachedCallStubKey& other) const
+    {
+        LIMITED_METHOD_CONTRACT;
+
+        if (HashCode != other.HashCode || NumRoutines != other.NumRoutines || TotalStackSize != other.TotalStackSize || Invoke != other.Invoke)
+            return false;
+
+        for (int i = 0; i < NumRoutines; i++)
+        {
+            if (Routines[i] != other.Routines[i])
+                return false;
+        }
+        return true;
+    }
+
+    const int32_t HashCode = 0;
+    const int NumRoutines = 0;
+    const int TotalStackSize = 0;
+    const CallStubHeader::InvokeFunctionPtr Invoke = NULL; // Pointer to the invoke function
+    const PCODE *Routines;
+};
+
+struct CachedCallStub
+{
+    CachedCallStub(int32_t hashCode, int numRoutines, PCODE *pRoutines, int totalStackSize, CallStubHeader::InvokeFunctionPtr pInvokeFunction) :
+        HashCode(hashCode),
+        Header(numRoutines, pRoutines, totalStackSize, pInvokeFunction)
+    {
+    }
+
+    int32_t HashCode;
+    CallStubHeader Header;
+
+    CachedCallStubKey GetKey()
+    {
+        return CachedCallStubKey(
+            HashCode,
+            Header.NumRoutines,
+            &Header.Routines[0],
+            Header.TotalStackSize,
+            Header.Invoke);
+    }
+
+    static COUNT_T Hash(const CachedCallStubKey& key)
+    {
+        LIMITED_METHOD_CONTRACT;
+        return key.HashCode;
+    }
+};
+
+static CrstStatic s_callStubCrst;
+
+typedef  PtrSHashTraits<CachedCallStub, CachedCallStubKey> CallStubCacheTraits;
+
+typedef SHash<CallStubCacheTraits> CallStubCacheHash;
+static CallStubCacheHash* s_callStubCache;
+
+void InitCallStubGenerator()
+{
+    STANDARD_VM_CONTRACT;
+
+    s_callStubCrst.Init(CrstCallStubCache);
+    s_callStubCache = new CallStubCacheHash;
+}
+
+CallStubHeader *CallStubGenerator::GenerateCallStubForSig(MetaSig &sig)
+{
+    STANDARD_VM_CONTRACT;
+
+    // Allocate space for the routines. The size of the array is conservatively set to twice the number of arguments
+    // plus one slot for the target pointer and reallocated to the real size at the end.
+    PCODE *pRoutines = (PCODE*)alloca(ComputeTempStorageSize(sig));
+
+    ComputeCallStub(sig, pRoutines);
```

Comment:
Couldn't we use the signature data to generate the hash instead of building the call stub and then throwing it away when we find it in the cache? It seems it defeats the purpose of caching the stub - we could then just generate it every time again and the perf would be the same.

Reply:
The store exists to handle the storage problem not the cost of calculating the callstub problem. Computing the callstub is actually very fast, but managing the storage is what we're doing here. Really I shouldn't call this a cache, but callings things like this caches is extremely prevalent in our codebase.